src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -2170,197 +2159,116 @@ public BigDecimal sqrt(MathContext mc) {
-            //
-            // conversion cycle, but it avoids the need for several
-            // Newton iterations in BigDecimal arithmetic to get the
-            // working answer to 15 digits of precision. If many fewer
-            // than 15 digits were needed, it might be faster to do
-            // the loop entirely in BigDecimal arithmetic.
-            //
-            // (A double value might have as many as 17 decimal
-            // digits of precision; it depends on the relative density
-            // of binary and decimal numbers at different regions of
-            // the number line.)
-            //
-            // (It would be possible to check for certain special
-            // cases to avoid doing any Newton iterations. For
-            // example, if the BigDecimal -> double conversion was
-            // known to be exact and the rounding mode had a
-            // low-enough precision, the post-Newton rounding logic
-            // could be applied directly.)
-
-            BigDecimal guess = new BigDecimal(Math.sqrt(working.doubleValue()));
-            int guessPrecision = 15;
-            int originalPrecision = mc.getPrecision();
-            int targetPrecision;
-
-            // If an exact value is requested, it must only need about
-            // half of the input digits to represent since multiplying
-            // an N digit number by itself yield a 2N-1 digit or 2N
-            // digit result.
-            if (originalPrecision == 0) {
-                targetPrecision = stripped.precision()/2 + 1;
-            } else {
-                /*
-                 * To avoid the need for post-Newton fix-up logic, in
-                 * the case of half-way rounding modes, double the
-                 * target precision so that the "2p + 2" property can
-                 * be relied on to accomplish the final rounding.
-                 */
-                switch (mc.getRoundingMode()) {
-                case HALF_UP:
-                case HALF_DOWN:
-                case HALF_EVEN:
-                    targetPrecision = 2 * originalPrecision;
-                    if (targetPrecision < 0) // Overflow
-                        targetPrecision = Integer.MAX_VALUE - 2;
-                    break;
-
-                default:
-                    targetPrecision = originalPrecision;
-                    break;
+                // After stripTrailingZeros, the representation is normalized as
+                //
+                // unscaledValue * 10^(-scale)
+                //
+                // where unscaledValue is an integer with the minimum
+                // precision for the cohort of the numerical value.
+                BigInteger working = stripped.unscaledValue();
+                final int resultScale;
+                if ((strippedScale & 1) == 0) {
+                    resultScale = strippedScale >> 1;
+                } else {
+                    working = working.multiply(10L);
```

Comment:
While this is correct, there's useless work that is being performed.

After the multiplication by 10, `working` cannot be an exact square, so this will fail later at L.2199.
I wonder if this can be simplified to avoid the `multiply()` and the following `sqrtAndRemainder()` when `strippedScale` is odd.

Reply:
I'm changing the code to take into account this fact.